core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp

```diff
@@ -128,21 +128,19 @@ inline void check_shmem_request(CudaInternal const* cuda_instance, int shmem) {
 // These functions need to be templated on DriverType and LaunchBounds
 // so that the static bool is unique for each type combo
 // KernelFuncPtr does not necessarily contain that type information.
-// FIXME_CUDA_MULTIPLE_DEVICES
 template <class DriverType, class LaunchBounds, class KernelFuncPtr>
 const cudaFuncAttributes& get_cuda_kernel_func_attributes(
-    const KernelFuncPtr& func) {
+    int cuda_device, const KernelFuncPtr& func) {
   // Only call cudaFuncGetAttributes once for each unique kernel
   // by leveraging static variable initialization rules
-  auto wrap_get_attributes = [&]() -> cudaFuncAttributes {
+  static std::map<int, cudaFuncAttributes> func_attr;
+  if (func_attr.find(cuda_device) == func_attr.end()) {
     cudaFuncAttributes attr;
-    KOKKOS_IMPL_CUDA_SAFE_CALL(
-        (CudaInternal::singleton().cuda_func_get_attributes_wrapper(&attr,
-                                                                    func)));
-    return attr;
-  };
-  static cudaFuncAttributes func_attr = wrap_get_attributes();
-  return func_attr;
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaSetDevice(cuda_device));
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaFuncGetAttributes(&attr, func));
+    func_attr.insert(cuda_device, attr);
+  }
+  return func_attr[cuda_device];
```

Comment:
If we are worried about latency we should do
```c++
auto search = func_attr.find(cuda_device);
if (search == func_attr.end()) {
  ...
  return attr;
}
return search->second;

Reply:
Running `Kokkos Launch Latency Test`:
This pull request
```
N V K M time_no_fence time_fence (time_no_fence_fenced)
100000000 1 1 20 parallel_for: 3.410850 3239.667250 ( 3237.208250 )
100000000 16 1 20 parallel_for: 3.512550 1010.526850 ( 1004.075800 )
100000000 200 1 20 parallel_for: 964.569050 1016.016200 ( 1014.913050 )
100000000 3000 1 20 parallel_for: 967.518000 1019.864300 ( 1017.874550 )
100000000 30000 1 20 parallel_for: 875.876850 1050.608750 ( 1028.366200 )
```
before
```
N V K M time_no_fence time_fence (time_no_fence_fenced)
100000000 1 1 20 parallel_for: 3.399300 3238.960950 ( 3237.724650 )
100000000 16 1 20 parallel_for: 3.511050 1011.017300 ( 1004.226550 )
100000000 200 1 20 parallel_for: 964.840100 1016.107300 ( 1015.046800 )
100000000 3000 1 20 parallel_for: 967.730900 1019.796150 ( 1018.087500 )
100000000 30000 1 20 parallel_for: 879.111450 1051.797000 ( 1031.848700 )
```
I don't see